File: utilslib/src/main/java/com/ashlikun/utils/ui/UiUtils.java
package com.ashlikun.utils.ui;

import android.app.Activity;
import android.content.Context;
import android.support.design.widget.TabLayout;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.ashlikun.utils.R;
import com.ashlikun.utils.other.DimensUtils;

import java.lang.reflect.Field;


public class UiUtils {

    public static void setColorSchemeResources(SwipeRefreshLayout swipeRefreshLayout) {
        swipeRefreshLayout.setColorSchemeResources(R.color.SwipeRefreshLayout_1, R.color.SwipeRefreshLayout_2, R.color.SwipeRefreshLayout_3, R.color.SwipeRefreshLayout_4);
    }

    public static void setRefreshing(final SwipeRefreshLayout swipeRefreshLayout, final boolean b) {
        swipeRefreshLayout.postDelayed(new Runnable() {
            @Override
            public void run() {
                swipeRefreshLayout.setRefreshing(b);
            }
        }, 400);

    }

    /**
     * 从资源文件获取一个view
     */
    public static View getInflaterView(Context context, int res) {
        View view = ((LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE)).inflate(res, null);
        return view;
    }

    /**
     * 从资源文件获取一个view
     */
    public static View getInflaterView(Context context, int res, ViewGroup parent) {
        View view = ((LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE)).inflate(res, parent);
        return view;
    }

    /**
     * 从资源文件获取一个view
     */
    public static View getInflaterView(Context context, int res, ViewGroup parent, boolean attachToRoot) {
        View view = ((LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE)).inflate(res, parent, attachToRoot);
        return view;
    }


    /**
     * 设置ImageView渲染（Tint）
     */
    public static void setImageViewTint(final ImageView view, final int color) {
        view.setColorFilter(view.getResources().getColor(color));
    }


    public static View getRootView(Activity context) {
        return context.findViewById(android.R.id.content);
    }

    public static View getDecorView(Activity context) {
        return context.getWindow().getDecorView();
    }


    /**
     * 按照原始的宽度，根据比例，缩放
     *
     * @param bili (w/h)
     */
    public static void scaleViewByWidth(final View view, final float bili) {
        getViewSize(view, new OnSizeListener() {
            @Override
            public void onSize(int width, int height) {
                scaleViewByWidth(view, width, bili);
            }
        });
    }

    /**
     * 同上
     *
     * @param view
     * @param width
     * @param bili
     */
    public static void scaleViewByWidth(View view, int width, final float bili) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        if (params == null) {
            params = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        }
        params.height = (int) (width / bili);
        view.setLayoutParams(params);
    }

    /**
     * 按照原始的高度，根据比例，缩放
     *
     * @param bili (w/h)
     */
    public static void scaleViewByHeight(final View view, final float bili) {
        getViewSize(view, new OnSizeListener() {
            @Override
            public void onSize(int width, int height) {
                scaleViewByHeight(view, height, bili);
            }
        });
    }

    /**
     * 同上
     *
     * @param view
     * @param height
     * @param bili
     */
    public static void scaleViewByHeight(final View view, int height, final float bili) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        if (params == null) {
            params = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        }
        params.width = (int) (height * bili);
        view.setLayoutParams(params);
    }

    /**
     * 设置view高度
     *
     * @param view
     * @param height
     */
    public static void setViewHeight(View view, int height) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        if (params == null) {
            params = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        }
        params.height = height;
        view.setLayoutParams(params);
    }

    /**
     * 设置view宽度
     *
     * @param view
     * @param width
     */
    public static void setViewWidth(View view, int width) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        if (params == null) {
            params = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        }
        params.width = width;
        view.setLayoutParams(params);
    }

    /**
     * 设置view大小
     *
     * @param view
     * @param width
     */
    public static void setViewSize(View view, int width, int height) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        if (params == null) {
            params = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        }
        params.width = width;
        params.height = height;
        view.setLayoutParams(params);
    }

    /**
     * 设置view   Margin
     *
     * @param view
     */
    public static void setViewMargin(View view, int leftMargin, int topMargin, int rightMargin, int bottomMargin) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        if (params != null && params instanceof ViewGroup.MarginLayoutParams) {
            ((ViewGroup.MarginLayoutParams) params).leftMargin = leftMargin;
            ((ViewGroup.MarginLayoutParams) params).topMargin = topMargin;
            ((ViewGroup.MarginLayoutParams) params).rightMargin = rightMargin;
            ((ViewGroup.MarginLayoutParams) params).bottomMargin = bottomMargin;
            view.setLayoutParams(params);
        }
    }

    /**
     * 获取view的Margin
     *
     * @param view
     * @param orientation 那个值，左上右下（0,1,2,3）
     * @return
     */
    public static int getViewMargin(View view, int orientation) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        if (params != null && params instanceof ViewGroup.MarginLayoutParams) {
            if (orientation == 0) {
                return ((ViewGroup.MarginLayoutParams) params).leftMargin;
            } else if (orientation == 1) {
                return ((ViewGroup.MarginLayoutParams) params).topMargin;
            } else if (orientation == 2) {
                return ((ViewGroup.MarginLayoutParams) params).rightMargin;
            } else if (orientation == 3) {
                return ((ViewGroup.MarginLayoutParams) params).bottomMargin;
            }
        }
        return -1;
    }

    /**
     * 获取view大小
     *
     * @param onSizeListener 监听回调
     */
    public static void getViewSize(final View view, final OnSizeListener onSizeListener) {
        ViewTreeObserver observer = view.getViewTreeObserver();
        observer.addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                if (view.getMeasuredHeight() <= 0 && view.getMeasuredWidth() <= 0) {
                    return;
                }
                view.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                onSizeListener.onSize(view.getMeasuredWidth(), view.getMeasuredHeight());
            }
        });
    }

    /**
     * 设置Tablayou的底部线宽度和文字一样宽，要在设置完item后设置
     *
     * @param tabLayout
     * @param marginDp  边距距离
     */
    public static void setTabLayoutLine(final TabLayout tabLayout, final int marginDp) {

        //了解源码得知 线的宽度是根据 tabView的宽度来设置的,post是为了可以获取到宽度
        tabLayout.post(new Runnable() {
            @Override
            public void run() {
                try {
                    //拿到tabLayout的mTabStrip属性
                    LinearLayout mTabStrip = (LinearLayout) tabLayout.getChildAt(0);
                    int margin = DimensUtils.dip2px(tabLayout.getContext(), marginDp);
                    for (int i = 0; i < mTabStrip.getChildCount(); i++) {
                        ViewGroup tabView = (ViewGroup) mTabStrip.getChildAt(i);
                        //拿到tabView的mTextView属性
                        TextView mTextView = null;
                        for (int j = 0; j < tabView.getChildCount(); j++) {
                            if (tabView.getChildAt(j) instanceof TextView) {
                                mTextView = (TextView) tabView.getChildAt(j);
                            }
                        }
                        //如果没有找到，就反射获取，一半不会
                        if (mTextView == null) {
                            Field mTextViewField = tabView.getClass().getDeclaredField("textView");
                            mTextViewField.setAccessible(true);
                            mTextView = (TextView) mTextViewField.get(tabView);
                            if (mTextView == null) {
                                Field mTextViewField2 = tabView.getClass().getDeclaredField("mTextView");
                                mTextViewField2.setAccessible(true);
                                mTextView = (TextView) mTextViewField2.get(tabView);
                            }
                        }
                        tabView.setPadding(0, 0, 0, 0);
                        //因为我想要的效果是   字多宽线就多宽，所以测量mTextView的宽度
                        int width = 0;
                        width = mTextView.getWidth();
                        if (width == 0) {
                            mTextView.measure(0, 0);
                            width = mTextView.getMeasuredWidth();
                        }
                        //设置tab左右间距  注意这里不能使用Padding 因为源码中线的宽度是根据 tabView的宽度来设置的
                        LinearLayout.LayoutParams params = (LinearLayout.LayoutParams) tabView.getLayoutParams();
                        params.width = width;
                        params.leftMargin = margin;
                        params.rightMargin = margin;
                        tabView.setLayoutParams(params);
                    }

                } catch (NoSuchFieldException e) {
                    e.printStackTrace();
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                }
            }
        });

    }
}

File: utilslib/src/main/java/com/ashlikun/utils/ui/UiUtils.kt
package com.ashlikun.utils.ui

import android.R
import android.animation.Animator
import android.animation.AnimatorSet
import android.animation.TimeInterpolator
import android.animation.ValueAnimator
import android.app.Activity
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.MarginLayoutParams
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.view.animation.LinearInterpolator
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.view.setPadding
import com.ashlikun.utils.other.DimensUtils.dip2px
import com.ashlikun.utils.ui.extend.getInflaterView
import com.ashlikun.utils.ui.extend.padding
import com.ashlikun.utils.ui.extend.setViewSize
import com.google.android.material.tabs.TabLayout

typealias OnSizeListener = (width: Int, height: Int) -> Unit

object UiUtils {
    /**
     * 从资源文件获取一个view
     * @param context 不同的 context 对应不同的LayoutInflater
     */
    fun getInflaterView(context: Context, res: Int): View {
        return LayoutInflater.from(context).inflate(res, null)
    }

    /**
     * 从资源文件获取一个view
     * @param context 不同的 context 对应不同的LayoutInflater
     */
    fun getInflaterView(context: Context, res: Int, parent: ViewGroup?): View {
        return LayoutInflater.from(context).inflate(res, parent)
    }

    /**
     * 从资源文件获取一个view
     * @param context 不同的 context 对应不同的LayoutInflater
     */
    fun getInflaterView(
        context: Context,
        res: Int,
        parent: ViewGroup?,
        attachToRoot: Boolean
    ): View {
        return LayoutInflater.from(context).inflate(res, parent, attachToRoot)
    }

    /**
     * 设置ImageView渲染（Tint）
     */
    fun setImageViewTint(view: ImageView, color: Int) {
        view.setColorFilter(view.resources.getColor(color))
    }

    @JvmStatic
    fun getRootView(context: Activity): View {
        return context.findViewById(R.id.content)
    }

    fun getDecorView(context: Activity): View {
        return context.window.decorView
    }

    /**
     * 按照原始的宽度，根据比例，缩放
     *
     * @param bili (w/h)
     */
    fun scaleViewByWidth(view: View, bili: Float) {
        getViewSize(view) { width, height -> scaleViewByWidth(view, width, bili) }
    }

    /**
     * 同上
     *
     * @param view
     * @param width
     * @param bili
     */
    fun scaleViewByWidth(view: View, width: Int, bili: Float) {
        var params = view.layoutParams
        if (params == null) {
            params = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        }
        params.height = (width / bili).toInt()
        view.layoutParams = params
    }

    /**
     * 按照原始的高度，根据比例，缩放
     *
     * @param bili (w/h)
     */
    fun scaleViewByHeight(view: View, bili: Float) {
        getViewSize(view) { width, height -> scaleViewByHeight(view, height, bili) }
    }

    /**
     * 同上
     *
     * @param view
     * @param height
     * @param bili
     */
    fun scaleViewByHeight(view: View, height: Int, bili: Float) {
        var params = view.layoutParams
        if (params == null) {
            params = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        }
        params.width = (height * bili).toInt()
        view.layoutParams = params
    }

    /**
     * 设置view高度
     *
     * @param view
     * @param height
     */
    fun setViewHeight(view: View, height: Int) {
        var params = view.layoutParams
        if (params == null) {
            params = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        }
        params.height = height
        view.layoutParams = params
    }

    /**
     * 设置view宽度
     *
     * @param view
     * @param width
     */
    fun setViewWidth(view: View, width: Int) {
        var params = view.layoutParams
        if (params == null) {
            params = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        }
        params.width = width
        view.layoutParams = params
    }

    /**
     * 设置view大小
     *
     * @param view
     * @param width
     */
    fun setViewSize(view: View, width: Int, height: Int) {
        view.setViewSize(width, height)
    }

    /**
     * 设置view   Margin
     * @param duration 动画执行时间 > 0 就启用动画
     * @param interpolator 动画的插值
     */
    fun setViewMargin(view: View,
                      leftMargin: Int? = null,
                      topMargin: Int? = null,
                      rightMargin: Int? = null,
                      bottomMargin: Int? = null,
                      duration: Long = 0,
                      interpolator: TimeInterpolator? = null) {
        val params = view.layoutParams
        if (params != null && params is MarginLayoutParams) {
            if (duration > 0) {
                val animSet = AnimatorSet()
                val animList = mutableListOf<Animator>()
                if (leftMargin != null) {
                    animList.add(ValueAnimator.ofInt(params.leftMargin, leftMargin).apply {
                        addUpdateListener {
                            val w = it.animatedValue as Int
                            params.leftMargin = w
                            view.layoutParams = params
                        }
                    })
                }
                if (topMargin != null) {
                    animList.add(ValueAnimator.ofInt(params.topMargin, topMargin).apply {
                        addUpdateListener {
                            val w = it.animatedValue as Int
                            params.topMargin = w
                            view.layoutParams = params
                        }
                    })
                }
                if (rightMargin != null) {
                    animList.add(ValueAnimator.ofInt(params.rightMargin, rightMargin).apply {
                        addUpdateListener {
                            val w = it.animatedValue as Int
                            params.rightMargin = w
                            view.layoutParams = params
                        }
                    })
                }
                if (bottomMargin != null) {
                    animList.add(ValueAnimator.ofInt(params.bottomMargin, bottomMargin).apply {
                        addUpdateListener {
                            val w = it.animatedValue as Int
                            params.bottomMargin = w
                            view.layoutParams = params
                        }
                    })
                }
                animSet.playTogether(animList)
                animSet.duration = duration
                animSet.interpolator = interpolator ?: LinearInterpolator()
                animSet.start()
            } else {
                if (leftMargin != null)
                    params.leftMargin = leftMargin
                if (topMargin != null)
                    params.topMargin = topMargin
                if (rightMargin != null)
                    params.rightMargin = rightMargin
                if (bottomMargin != null)
                    params.bottomMargin = bottomMargin
                view.layoutParams = params
            }

        }
    }

    /**
     * 设置view   Padding
     * @param duration 动画执行时间 > 0 就启用动画
     * @param interpolator 动画的插值
     */
    fun setPaddings(view: View,
                    padding: Int? = null,
                    leftPadding: Int? = null,
                    topPadding: Int? = null,
                    rightPadding: Int? = null,
                    bottomPadding: Int? = null,
                    duration: Long = 0,
                    interpolator: TimeInterpolator? = null) {
        if (duration > 0) {
            val animSet = AnimatorSet()
            val animList = mutableListOf<Animator>()
            if (padding != null) {
                animList.add(ValueAnimator.ofInt(view.padding(), padding).apply {
                    addUpdateListener {
                        view.setPadding(padding)
                    }
                })
            } else {
                if (leftPadding != null) {
                    animList.add(ValueAnimator.ofInt(view.paddingLeft, leftPadding).apply {
                        addUpdateListener {
                            view.setPadding(leftPadding, view.paddingTop, view.paddingRight, view.paddingBottom)
                        }
                    })
                }
                if (topPadding != null) {
                    animList.add(ValueAnimator.ofInt(view.paddingTop, topPadding).apply {
                        addUpdateListener {
                            view.setPadding(view.paddingLeft, topPadding, view.paddingRight, view.paddingBottom)
                        }
                    })
                }
                if (rightPadding != null) {
                    animList.add(ValueAnimator.ofInt(view.paddingRight, rightPadding).apply {
                        addUpdateListener {
                            view.setPadding(view.paddingLeft, view.paddingTop, rightPadding, view.paddingBottom)
                        }
                    })
                }
                if (bottomPadding != null) {
                    animList.add(ValueAnimator.ofInt(view.paddingBottom, bottomPadding).apply {
                        addUpdateListener {
                            view.setPadding(view.paddingLeft, view.paddingTop, view.paddingRight, bottomPadding)
                        }
                    })
                }
            }
            animSet.playTogether(animList)
            animSet.duration = duration
            animSet.interpolator = interpolator ?: LinearInterpolator()
            animSet.start()
        } else {
            if (padding != null) {
                view.setPadding(padding)
            } else {
                view.setPadding(leftPadding ?: view.paddingLeft,
                    topPadding ?: view.paddingTop,
                    rightPadding ?: view.paddingRight,
                    bottomPadding ?: view.paddingBottom)
            }
        }

    }

    /**
     * 获取view的Margin
     *
     * @param view
     * @param orientation 那个值，左上右下（0,1,2,3）
     * @return
     */
    fun getViewMargin(view: View, orientation: Int): Int {
        val params = view.layoutParams
        if (params != null && params is MarginLayoutParams) {
            if (orientation == 0) {
                return params.leftMargin
            } else if (orientation == 1) {
                return params.topMargin
            } else if (orientation == 2) {
                return params.rightMargin
            } else if (orientation == 3) {
                return params.bottomMargin
            }
        }
        return -1
    }

    /**
     * 获取view大小
     */
    fun getViewSize(view: View, onSizeListener: OnSizeListener? = null) {
        if (view.measuredWidth > 0 || view.measuredHeight > 0) {
            onSizeListener?.invoke(view.measuredWidth, view.measuredHeight)
            return
        }
        val observer = view.viewTreeObserver
        observer.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                if (view.measuredHeight <= 0 && view.measuredWidth <= 0) {
                    return
                }
                view.viewTreeObserver.removeOnGlobalLayoutListener(this)
                onSizeListener?.invoke(view.measuredWidth, view.measuredHeight)
            }
        })
    }

    /**
     * 设置Tablayou的底部线宽度和文字一样宽，要在设置完item后设置
     *
     * @param tabLayout
     * @param marginDp  边距距离
     */
    fun setTabLayoutLine(tabLayout: TabLayout, marginDp: Int) {

        //了解源码得知 线的宽度是根据 tabView的宽度来设置的,post是为了可以获取到宽度
        tabLayout.post {
            try {
                //拿到tabLayout的mTabStrip属性
                val mTabStrip = tabLayout.getChildAt(0) as LinearLayout
                val margin = dip2px(tabLayout.context, marginDp.toFloat())
                for (i in 0 until mTabStrip.childCount) {
                    val tabView = mTabStrip.getChildAt(i) as ViewGroup
                    //拿到tabView的mTextView属性
                    var mTextView: TextView? = null
                    for (j in 0 until tabView.childCount) {
                        if (tabView.getChildAt(j) is TextView) {
                            mTextView = tabView.getChildAt(j) as TextView
                        }
                    }
                    //如果没有找到，就反射获取，一半不会
                    if (mTextView == null) {
                        val mTextViewField = tabView.javaClass.getDeclaredField("textView")
                        mTextViewField.isAccessible = true
                        mTextView = mTextViewField[tabView] as TextView
                        if (mTextView == null) {
                            val mTextViewField2 = tabView.javaClass.getDeclaredField("mTextView")
                            mTextViewField2.isAccessible = true
                            mTextView = mTextViewField2[tabView] as TextView
                        }
                    }
                    tabView.setPadding(0, 0, 0, 0)
                    //因为我想要的效果是   字多宽线就多宽，所以测量mTextView的宽度
                    var width = 0
                    width = mTextView!!.width
                    if (width == 0) {
                        mTextView.measure(0, 0)
                        width = mTextView.measuredWidth
                    }
                    //设置tab左右间距  注意这里不能使用Padding 因为源码中线的宽度是根据 tabView的宽度来设置的
                    val params = tabView.layoutParams as LinearLayout.LayoutParams
                    params.width = width
                    params.leftMargin = margin
                    params.rightMargin = margin
                    tabView.layoutParams = params
                }
            } catch (e: NoSuchFieldException) {
                e.printStackTrace()
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }
        }
    }
}